internal/retry/request.go

```diff
@@ -0,0 +1,248 @@
+	switch s {
+	case NoRetry:
+		return "NoRetry"
+	case RetryImmediate:
+		return "RetryImmediate"
+	case RetryWithInitialDelay:
+		return "RetryWithInitialDelay"
+	default:
+		panic("Invalid RetryStrategy")
+	}
+}
+
+// A RetryStrategyFn is used to determine whether or not a retry is needed, and
+// if so, whether an initial delay is needed or not.
+type RetryStrategyFn func(*http.Response) (RetryStrategy, error)
+
+// A RetryAfterFn is used to parse the delay from the Retry-After header.
+//
+// The returned time.Duration will be used as the delay.
+//
+// If the header is not set, or the value cannot be parsed, a Duration of 0
+// must be returned.
+type RetryAfterFn func(*http.Response) time.Duration
+
+// A BackoffFn takes the previous delay and calculates a new, longer delay.
+//
+// This is used to extend the time between retries.
+type BackoffFn func(time.Duration) time.Duration
+
+// SleepFn must cause the current goroutine to sleep for the allocated Duration.
+//
+// The usual implementation of this is time.Sleep. This is provided for testing.
+type sleepFn func(time.Duration)
+
+// DefaultBackoff will double the duration d if it is greater than zero,
+// otherwise it returns 1 minute.
+func DefaultBackoff(d time.Duration) time.Duration {
+	if d <= 0 {
+		return time.Minute
+	} else {
+		return d * 2
+	}
+}
+
+type Options struct {
+	maxRetries         int
+	initialDelay       time.Duration
+	backoff            BackoffFn
+	sleep              sleepFn
+	retryAfter         RetryAfterFn
+	retryStrategyFuncs []RetryStrategyFn
+}
+type Option interface {
+	Apply(*Options)
+}
+type optionFn func(*Options)
+
+func (o optionFn) Apply(opts *Options) {
+	o(opts)
+}
+
+func MaxRetries(max int) Option {
+	return optionFn(func(o *Options) {
+		o.maxRetries = max
+	})
+}
+
+func Backoff(bo BackoffFn) Option {
+	return optionFn(func(o *Options) {
+		o.backoff = bo
+	})
+}
+
+func InitialDelay(d time.Duration) Option {
+	return optionFn(func(o *Options) {
+		o.initialDelay = d
+	})
+}
+
+func RetryAfter(ra RetryAfterFn) Option {
+	return optionFn(func(o *Options) {
+		o.retryAfter = ra
+	})
+}
+
+func Strategy(rs RetryStrategyFn) Option {
+	return optionFn(func(o *Options) {
+		o.retryStrategyFuncs = append(o.retryStrategyFuncs, rs)
+	})
+}
+
+func MakeOptions(os ...Option) *Options {
+	opts := &Options{
+		maxRetries:   DefaultMaxRetries,
+		initialDelay: DefaultInitialDuration,
+		sleep:        time.Sleep,
+		backoff:      DefaultBackoff,
+	}
+	for _, o := range os {
+		o.Apply(opts)
+	}
+	return opts
+}
+
+type Request struct {
+	client   func(*http.Request) (*http.Response, error)
+	r        *http.Request
+	o        *Options
+	attempts int
+	done     bool
+	delay    time.Duration
+}
+
+func NewRequest(r *http.Request, client func(*http.Request) (*http.Response, error), o *Options) *Request {
+	return &Request{
+		client: client,
+		r:      r,
+		o:      o,
+	}
+}
+
+func (r *Request) Done() bool {
+	return r.done || r.attempts >= r.o.maxRetries
```

Comment:
does `attempts` include the first attempt? If so, semantically this should be `attempts > maxRetries` right?

Reply:
Semantically yes. And I will change this to match.

This means `Options.maxRetries == 0` always results in a single attempt, but no retry.